Lineage documentation updates (#1910)
astronomer/docs: astro/set-up-data-lineage.md

This guide provides information about how lineage data is automatically extracted from Apache Airflow tasks on Astro and how to integrate external systems, including Databricks and dbt, that require additional configuration. To learn about how to view data lineage on Astro, see [View data lineage](data-lineage.md).

> ## Extract lineage data from Airflow operators using supported extractors

Reviewer: This is a comment that applies throughout the corpus of documentation on lineage -

I typically call it "lineage metadata", since it's data about data. I don't know if this is a worthwhile distinction, but I choose to do it to reinforce the notion that lineage run state updates don't contain any of the data itself.

Author: Good call, I'll defer to the OL docs for this style choice. I always thought "lineage metadata" would consist of the stats we surface in the UI, e.g. "Number of runs"/ "Last run duration"/ etc, while lineage data was what the specific data the run contained.

The latest version of the OpenLineage Airflow library is installed on [Astro Runtime](runtime-image-architecture.md) by default, meaning that you can use all default extractors without additional configuration. If you use an Airflow operator that includes a default extractor in your DAG, the operator automatically generates lineage data to the **Lineage** page on Astro. 

Each operator generates different lineage data based on its default extractor. For more information about operators with default extractors and what lineage data they generate, see [OpenLineage documentation](https://openlineage.io/docs/integrations/about#capability-matrix).

## Extract lineage data from Airflow operators using custom extractors

If you want to extract lineage data from an Airflow operator that doesn't have a default extractor, you can write a custom extractor and add it to your Astro project.

> To write a custom extractor, see [OpenLineage documentation](https://openlineage.io/docs/integrations/airflow/extractors/custom-extractors). To add a custom extractor to an Astro Deployment:

Reviewer 2: ```suggestion
To write a custom extractor, see [OpenLineage documentation](https://openlineage.io/docs/integrations/airflow/extractors/custom-extractors).
```

1. Add your custom extractor files to the `include` folder of your local Astro project.
2. Deploy your project. See [Deploy code](deploy-code.md).

> 3. Set the following [environment variable](environment-variables.md) in your Astro Deployment:

Reviewer 2: Do you want to link to the topic that describes how to add an environment variable?

Author: The linked doc describes a few different ways of doing this, so I'm okay linking generally and letting the user decide where they want to set it.

    ```text

>     OPENLINEAGE_EXTRACTORS='<path-to-extractor-class-1>;<path-to-extractor-class-2>;<path-to-extractor-class-x>

Reviewer: I wonder if we should specify whether it should be specified as a relative path or an absolute one?

    ```

    If you are importing only one custom extractor, do not include a semicolon after the file path. 

> ## Extract lineage data from Airflow operators using custom inlets and outlets

Author: Is this really "Lineage data"? Is this an alternative to facets/ extractors, or in addition to them?

Reviewer: This is an alternative to extractors, but it's still lineage metadata.

(Facets are orthogonal - I imagine the manual method still makes use of some facets.)

> An alternative to writing a custom extractor is to specify dataset inlets and outlets directly in your task parameters. These inlets and outlets will render as dependency lines on the lineage graph for the DAG. This option is suitable if your main priority is rendering an accurate lineage graph of your DAG, and you don't need to generate specific facets from your operators.

Author: ```suggestion
An alternative to writing a custom extractor is to specify dataset inlets and outlets directly in your task parameters. These inlets and outlets appear as dependency lines on the lineage graph for the DAG. This option is suitable if your main priority is rendering an accurate lineage graph of your DAG, and you don't need to generate specific facets from your operators.
```

> To specify inlets and outlets, see the [OpenLineage documentation](https://openlineage.io/docs/integrations/airflow/manual) and [Apache Airflow documentation](https://airflow.apache.org/docs/apache-airflow/stable/administration-and-deployment/lineage.html). Note that OpenLineage only supports specifying inlets and outlets using `Table` objects.

Reviewer 2: ```suggestion
To specify inlets and outlets, see the [OpenLineage documentation](https://openlineage.io/docs/integrations/airflow/manual) and [Apache Airflow documentation](https://airflow.apache.org/docs/apache-airflow/stable/administration-and-deployment/lineage.html). OpenLineage only supports specifying inlets and outlets using `Table` objects.
```

## Extract lineage data from external systems to Astro

> To send lineage data integrate from an external system to Astro, you need to configure the external system's OpenLineage integration with a Deployment namespace, your Organization's OpenLineage URL, and your organization's OpenLineage API key. This information is used to send OpenLineage data to your Astro lineage backend.

Reviewer 2: ```suggestion
To send lineage data from an external system to Astro, you need to configure the external system's OpenLineage integration with a Deployment namespace, your Organization's OpenLineage URL, and your organization's OpenLineage API key. This information is used to send OpenLineage data to your Astro lineage backend.
```

> To locate your Deployment namespace in the Cloud UI, select a Workspace and then copy the value with the format `<text>-<text>-<four-digit-number>` next to the Deployment name. To locate your Organization's OpenLineage URL and OpenLineage API key, go to `https://cloud.<your-astro-base-domain>.io/settings` and copy the values in the **Lineage API Key** and **OpenLineage URL** fields.

Reviewer 2: ```suggestion
To locate your Deployment namespace in the Cloud UI, select a Workspace and then copy the value with the format `<text>-<text>-<four-digit-number>` next to the Deployment name. To locate your Organization's OpenLineage URL and OpenLineage API key, go to `https://cloud.<your-astro-base-domain>.io/settings` and copy the values in the **Lineage API Key** and **OpenLineage URL** fields.
```
It's unclear where the Deployment namespace value is located.

> Use the following topics to configure these values in supported external systems and send lineage data from those systems to Astro.

Reviewer 2: Is this statement necessary? Are users reading documentation linearly?

- Value: `False`

Astronomer recommends enabling this feature only for Deployments with non-sensitive code. For more information about Workspace permissions, see [Workspace roles](user-permissions.md#workspace-roles).

> ## Generate custom facets for OpenLineage events

Author: AFAIK the only way to generate custom facets for Airflow is through extractors. Is this true, and if so should we be more explicit about that?

Reviewer: I believe that is true for Airflow; generating custom facets in other systems will vary. It is also possible to send lineage metadata with facets using the `openlineage-python` client.
